### PR TITLE
[Rewriter] Move `nsys_profile.py` from onnxrewirter to onnxscript

### PR DESCRIPTION
**Pull request by titaiwangms**
1. Add `nsys_profile.py`
2. Add `ort_rewrite.py` (adapted from `post_process.py`)
3. Add README.md to elaborate on optimization development
4. DCE is required to be done after rewriting patterns: https://github.com/microsoft/onnx-rewriter/blob/78f678975b2f559ba9fc038a7b6be75adb89da7f/onnxrewriter/rewriter/onnxruntime/__init__.py#L55-L58

### CODE REVIEW

File: examples/bench_model.py
Comment (justinchuby) [inline examples/bench_model.py:1]: Move them to under tools/ ? It doesn't look like these are user facing examples?

File: onnxscript/rewriter/__init__.py
Comment (justinchuby) [inline onnxscript/rewriter/__init__.py:1]: Should this change belong to another pr?

File: examples/bench_model.py
Comment (titaiwangms) [reply examples/bench_model.py:1]: Fine to me

File: onnxscript/rewriter/__init__.py
Comment (titaiwangms) [reply onnxscript/rewriter/__init__.py:1]: I think it's fine. But if others find this is important to be in a stand alone PR, I can do that.

File: examples/bench_model.py
Comment (titaiwangms) [reply examples/bench_model.py:1]: Done

File: tools/README.md
Comment (BowenBao) [inline tools/README.md:1]: Maybe putting them under a dedicated folder under tools, since the theme is around benchmarking and profiling performance for optimization and rewrite.

File: tools/README.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+1. Setup access to vm with GPU.
+2. Install PyTorch, ONNX, ONNXScript, onnx-script, etc.
+3. Build ONNXRuntime from source, with nvtx profiling enabled.
+    ```
+    # Under onnxruntime root folder
+    ./build.sh --config RelWithDebInfo --parallel 0 --use_cuda --build_wheel --skip_tests --enable_nvtx_profile
+    # After build complete
+    pip install build/Linux/RelWithDebInfo/dist/onnxruntime_gpu-*.whl
+    ```
+<!-- 3. (Optional) Install torchbenchmark and related dependencies. -->
```
Comment (BowenBao) [inline tools/README.md:14]: remove comments xd

File: tools/README.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
+## Workflow
+
+### Setup Environment
+
+1. Setup access to vm with GPU.
+2. Install PyTorch, ONNX, ONNXScript, onnx-script, etc.
```
Comment (BowenBao) [inline tools/README.md:6]:
```suggestion
2. Install PyTorch, ONNX, ONNXScript, etc.
```

File: tools/README.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        │   ├── <model_name>_dynamo.onnx
+        │   ├── test_data_set_0
+        ├── torchscript
+        │   ├── <model_name>_torchscript.onnx
+        │   ├── test_data_set_0
+        ```
+
+3. Run optimization. Example command to apply optimizations on `stable_diffusion_unet` `dynamo` model and produces `dynamo_ort_rewritten` model under the same folder. `CUDA_VISIBLE_DEVICES` is used to specify GPU device to avoid the device others are using.
+    ```
+    CUDA_VISIBLE_DEVICES="3" python ort_rewrite.py.py  --model stable_diffusion_unet --model-dir ./onnx_models/ --log-level 20 --fusion
```
Comment (BowenBao) [inline tools/README.md:36]:
```suggestion
    CUDA_VISIBLE_DEVICES="3" python ort_rewrite.py  --model stable_diffusion_unet --model-dir ./onnx_models/ --log-level 20 --fusion
```

File: tools/README.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        CUDA_VISIBLE_DEVICES="2" python bench_model.py --model-dir ./onnx_models/stable_diffusion_unet --device cuda -i 20
+        ```
+    - Load and inspect model in Netron.
+        - Load `dynamo` model to inspect exported structured graph (optimizer already included).
+        - Load `dynamo_ort_rewritten` model to inspect inlined graph after onnxruntime rewriter.
+
+5. Develop optimization code.
+    - `onnx-script/onnxscript/optimizer`: Optimizations such as constant folding, inlining, dead code elimination etc.
+    - `onnx-script/onnxscript/rewriter`: Pattern based fusions.
+    - `onnx-script/onnxscript/rewriter/functions`: Function based fusions.
```
Comment (BowenBao) [inline tools/README.md:132]: This is now outdated with the change of directories.

**Review by BowenBao — COMMENTED** [review]
LG w/ comments.

Lintrunner is failing and will block merge. You'd need to either fix them if they are trivial, or add the files to lintrunner ignore list.

File: tools/nsys_profile.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+import argparse
+import datetime
+import os
+import subprocess
+
+import profile_analysis
+
+
+def nsys_profile_benchmark(
+    compiler: str, model_dir: str, iteration: int, report_name: str, nsys_path: str
```
Comment (BowenBao) [inline tools/nsys_profile.py:23]: Was confused about this PYLINT message for a bit, but it seems we should create a `main` function and move everything under `if __name__ == '__main__'` block there.

**Review by BowenBao — APPROVED** [review]
Forgot the stamp in previous review comment.

**Comment by justinchuby** [comment]
> LG w/ comments.
> 
> Lintrunner is failing and will block merge. You'd need to either fix them if they are trivial, or add the files to lintrunner ignore list.

Better fix it now than later

**Comment by justinchuby** [comment]
To fix the redefinition of names error, like Bowen suggested, I would

1. Create a `main()` function and move the logic from the `if __name__ == "__main__"` magic into it
2. Call `main()` from `if __name__ == "__main__"`

**Review by justinchuby — APPROVED** [review]
